lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return "", err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	var hostID string
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		if !api.StatusErrorCheck(err, http.StatusNotFound) {
+			return "", err
+		}
+
+		hostname, err := d.getHostname()
+		if err != nil {
+			return "", err
+		}
+
+		hostID, err = d.powerFlex().createHost(hostname, nqn)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if hostID == "" {
+		hostID = host.ID
+	}
+
+	return hostID, nil
+}
+
+// deleteNVMeHost delets this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) deleteNVMeHost() error {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		// Skip the deletion if the host doesn't exist anymore.
+		if api.StatusErrorCheck(err, http.StatusNotFound) {
+			return nil
+		}
+
+		return err
+	}
+
+	return d.powerFlex().deleteHost(host.ID)
+}
+
+// mapNVMeVolume maps the given volume onto this host.
+func (d *powerflex) mapNVMeVolume(name string) (revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	hostID, err := d.createNVMeHost()
+	if err != nil {
+		return nil, err
+	}
+
+	volumeID, err := d.powerFlex().getVolumeID(name)
+	if err != nil {
+		return nil, err
+	}
+
+	volume, err := d.powerFlex().getVolume(volumeID)
+	if err != nil {
+		return nil, err
+	}
+
+	mapped := false
+	for _, mapping := range volume.MappedSdcInfo {
+		if mapping.SdcID == hostID {
+			mapped = true
+		}
+	}
+
+	if !mapped {
+		err = d.powerFlex().createHostVolumeMapping(hostID, volumeID)
+		if err != nil {
+			return nil, err
+		}
+
+		revert.Add(func() { _ = d.powerFlex().deleteHostVolumeMapping(hostID, volumeID) })
+	}
+
+	cleanup := revert.Clone().Fail
+	revert.Success()
+	return cleanup, nil
+}
+
+// getNVMeMappedDevPath returns the local device path for the given NVMe volume name.
+// Set mapVolume to true if the volume isn't already mapped to this host.
+func (d *powerflex) getNVMeMappedDevPath(name string, mapVolume bool) (string, revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	if mapVolume {
+		cleanup, err := d.mapNVMeVolume(name)
+		if err != nil {
+			return "", nil, err
+		}
+
+		revert.Add(cleanup)
+
+		// Connect to the NVMe/TCP subsystem.
+		// We have to connect after the first mapping was established.
+		// PowerFlex does not offer any discovery log entries until a volume gets mapped to the host.
+		// This action is idempotent.
+		err = d.connectNVMeSubsys()
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	powerFlexVolumes := make(map[string]string)
+	discoverFunc := func(volumeID string) error {
+		basePath := "/dev/disk/by-id"
```

Comment:
We have devDiskByID constant for this btw

Reply:
Would it be feasible to have some logic over in the `resources` package together with the constant for looking up the drives? Something like `GetDisksById(filter)`.

Reply:
Added in a new `func GetDisksByID(filterPrefix string) []string` for general use.